src/pudl/transform/nrelatb.py

```diff
@@ -0,0 +1,415 @@
+"""Transform NREL ATB data into well normalized, cleaned tables."""
+
+from typing import Self
+
+import pandas as pd
+from dagster import asset
+from pydantic import BaseModel, field_validator
+
+import pudl.helpers as helpers
+
+IDX_ALL = [
+    "report_year",
+    "core_metric_case",
+    "core_metric_variable_year",
+    "cost_recovery_period_years",
+    "scenario_atb",
+    "technology_description",
+    "technology_description_detail_1",
+    "technology_description_detail_2",
+    "core_metric_parameter",
+]
+"""Expected primary key columns for the raw nrelatb data.
+
+The normalized core tables we are trying to build will result in a subset of these columns.
+"""
+
+
+class TableNormalizer(BaseModel):
+    """Info needed to convert a selection of the raw NREL table into a normalized table."""
+
+    idx: list[str]
+    """Primary key columns of normalized subset table."""
+    columns: list[str]
+    """Columns reported in raw NREL table that are unique to the :attr:`idx`."""
+
+
+def transform_normalize(nrelatb: pd.DataFrame, normalizer: TableNormalizer):
+    """Normalize a subset of the NREL ATB data into a small table.
+
+    Given a :class:`TableNormalizer` with a set of primary keys (``idx``) and a
+    list of columns (``columns``), build a table with just those primary key and
+    data columns from the larger ATB semi-processed table (output of
+    :func:`_core_nrelatb__transform_start`). Ensure that the output table is
+    unquie absed on the primary keys.
+    """
+    smol_table = nrelatb[normalizer.idx + normalizer.columns].drop_duplicates()
+    if not (
+        idx_dupes := smol_table[smol_table.duplicated(normalizer.idx, keep=False)]
+    ).empty:
+        raise AssertionError(
+            f"Found duplicate records on expected primary keys ({normalizer.idx}) and"
+            f"expected none. Duplicates:\n{idx_dupes}"
+        )
+    # drop the fully null columns bc some years there are no values in these
+    # columns and there is no use keeping these fully null records
+    return smol_table.dropna(how="all", subset=normalizer.columns)
+
+
+class Normalizer(BaseModel):
+    """Class that defines how to normalize all of the NREL tables that get the :func:`transform_normalize` treatment.
+
+    There are several columns in the raw ATB data that are not a part of the primary keys
+    for the tables that get the :func:`transform_unstack` treatment. This class helps us
+    build these smaller tables which have a smaller subset of primary key columns.
+    """
+
+    # There is info about revisions in the old data, but not in the 2021 -> current data
+    # revisions: TableNormalizer = TableNormalizer(
+    #     idx=["report_year"],
+    #     columns=["revision_num", "update_date"],
+    # )
```

Comment:
it probably makes more sense to take this out entirely. maybe i'll add this context into the issue that i made which would enable the older years of data.